### PR TITLE
Draft #1: Use HostExecutionContextManager to flow culture across threads.

### PR DESCRIPTION
**Pull request by troydai**
Disclaimer: this is a draft. All the traces printing will be removed and the culture could be placed in a better structured class. The naming are temporarily please suggest a more suitable name.

The changes successfully flow the culture across the threads in a Console application. Test application is here: https://github.com/troydai/dnx-issue1211

@davidfowl @DamianEdwards @GrabYourPitchforks

### CODE REVIEW

File: src/dnx.clr.managed/AspNetHostExecutionContextManager.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.Threading;
+
+namespace dnx.clr.managed
+{
+    public class AspNetHostExecutionContextManager : HostExecutionContextManager
```
Comment (troydai) [inline src/dnx.clr.managed/AspNetHostExecutionContextManager.cs]: Will be `internal`

**Comment by troydai** [comment]
An execution log:
```
[AspNetHostExecutionContextManager] Capture 1|en-US
[AspNetHostExecutionContextManager] Set culture of Context 0 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 2 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 4 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 6 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 6 from en-US to en-US
[AspNetHostExecutionContextManager] Revert culture of Context 4 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 2 from en-US to en-US
[AspNetHostExecutionContextManager] Revert culture of Context 0 from en-US to en-US
[AspNetHostExecutionContextManager] Capture 1|en-US
[AspNetHostExecutionContextManager] Set culture of Context 9 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 11 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 11 from en-US to en-US
[AspNetHostExecutionContextManager] Capture 1|en-US
[AspNetHostExecutionContextManager] Set culture of Context 14 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 16 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 16 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 14 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 9 from en-US to en-US
[AspNetHostExecutionContextManager] Capture 1|en-US
[AspNetHostExecutionContextManager] Set culture of Context 19 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 21 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 21 from en-US to en-US
[AspNetHostExecutionContextManager] Capture 1|en-US
[AspNetHostExecutionContextManager] Set culture of Context 24 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 26 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 26 from en-US to en-US
[AspNetHostExecutionContextManager] Capture 1|en-US
[AspNetHostExecutionContextManager] Set culture of Context 29 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 31 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 33 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 33 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 31 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 29 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 24 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 19 from en-US to en-US
[AspNetHostExecutionContextManager] Capture 1|en-US
[AspNetHostExecutionContextManager] Set culture of Context 36 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 36 from en-US to en-US
[AspNetHostExecutionContextManager] Capture 1|en-US
[AspNetHostExecutionContextManager] Set culture of Context 37 from en-US to en-US
[AspNetHostExecutionContextManager] Capture 7|en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 37 from en-US to en-US
[AspNetHostExecutionContextManager] Set culture of Context 38 from en-US to en-US
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 38 from en-US to en-US
1|zh-CN
[AspNetHostExecutionContextManager] Capture 1|zh-CN
[AspNetHostExecutionContextManager] Capture 1|zh-CN
[AspNetHostExecutionContextManager] Capture 1|zh-CN
[AspNetHostExecutionContextManager] Capture 1|zh-CN
[AspNetHostExecutionContextManager] Capture 1|zh-CN
[AspNetHostExecutionContextManager] Set culture of Context 44 from en-US to zh-CN
[AspNetHostExecutionContextManager] Set culture of Context 40 from zh-CN to zh-CN
4|zh-CN
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 40 from zh-CN to zh-CN
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 44 from zh-CN to en-US
[AspNetHostExecutionContextManager] Set culture of Context 41 from en-US to zh-CN
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 41 from zh-CN to en-US
[AspNetHostExecutionContextManager] Set culture of Context 42 from en-US to zh-CN
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 42 from zh-CN to en-US
[AspNetHostExecutionContextManager] Set culture of Context 43 from en-US to zh-CN
[AspNetHostExecutionContextManager] Revert - run revert action
[AspNetHostExecutionContextManager] Revert culture of Context 43 from zh-CN to en-US
```

**Comment by troydai** [comment]
@Praburaj

File: src/dnx.clr.managed/DomainManager.cs [hunk 2]
```diff
@@ -7,10 +7,12 @@
 using System.Threading;
 using dnx.host;
 using Microsoft.Framework.Runtime;
+using dnx.clr.managed;
```
Comment (Praburaj) [inline src/dnx.clr.managed/DomainManager.cs]: sort namespaces.

File: src/dnx.clr.managed/AspNetHostExecutionContextManager.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                Console.WriteLine("[{0}] Revert - call base", GetType().Name);
+                base.Revert(previousState);
+            }
+        }
+
+        private class AspNetHostExecutionContext : HostExecutionContext
+        {
+            // TODO: Remove debug only
+            private static int s_accumlate = 0;
+            public int _id = s_accumlate++;
```
Comment (Praburaj) [inline src/dnx.clr.managed/AspNetHostExecutionContextManager.cs]: Should this has an interlocked increment?

Comment (troydai) [inline src/dnx.clr.managed/AspNetHostExecutionContextManager.cs]: These two variables will be removed.

**Comment by troydai** [comment]
Further testing:
1) Under Helios;
2) Create child AppDomain.

**Comment by davidfowl** [comment]
@GrabYourPitchforks can you take a look at this?

**Comment by troydai** [comment]
Make it work in Helios: https://github.com/aspnet/Helios/pull/121

File: src/dnx.clr.managed/DnxHostExecutionContextManager.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            if (castHostExecutionContext != null)
+            {
+                object baseRevertParameter = null;
+                if (castHostExecutionContext.BaseContext != null)
+                {
+                    baseRevertParameter = base.SetHostExecutionContext(castHostExecutionContext.BaseContext);
+                }
+
+                var originalCulture = Thread.CurrentThread.CurrentCulture;
+                Thread.CurrentThread.CurrentCulture = castHostExecutionContext.ClientCulture;
```
Comment (DamianEdwards) [inline src/dnx.clr.managed/DnxHostExecutionContextManager.cs]: Need to set `Thread.CurrentThread.CurrentUICulture` too

Comment (troydai) [inline src/dnx.clr.managed/DnxHostExecutionContextManager.cs]: Updating ...

**Comment by troydai** [comment]
Updated :watch: 
@davidfowl @DamianEdwards @GrabYourPitchforks

**Comment by troydai** [comment]
@davidfowl did a local throughput test using ab. Doesn't observe regressing in terms of request per second.

**Comment by davidfowl** [comment]
What about cold start?

**Comment by troydai** [comment]
No regression observed in cold start, either. Tests are done on IIS and IIS Express. Took 10 samples.

**Comment by davidfowl** [comment]
:shipit:

**Comment by troydai** [comment]
Yo.

**Pull request merged**